### PR TITLE
Adding ggnn_encoder to support graph-to-sequence processing with attention and copy mechanism

### PR DESCRIPTION
**Pull request by SteveKommrusch**
I have added the GGNN network based on github.com/JamesChuanggg/ggnn.pytorch.git which is based on the paper "Gated Graph Sequence Neural Networks" by Y. Li, D. Tarlow, M. Brockschmidt, and R. Zemel.

We used this network for program equivalence proof generation as per our ArXiv paper "Equivalence of Dataflow Graphs via Rewrite Rules Using a Graph-to-Sequence Neural Model" at: https://arxiv.org/abs/2002.06799 . That paper shows the benefit of the graph-to-sequence model over a sequence-to-sequence model for our problem. My integration of the network into OpenNMT-py supports attention on the nodes as well as a copy mechanism.

### CODE REVIEW

**Comment by vince62s** [comment]
Hi @SteveKommrusch 
This is interesting even though I don't really see the use case myself.
A few comments:
This would require a specific .md file in docs/source (look at the index to understand the way it's done)
You are not supposed to change the pull_request_chk file, instead you would need to adapt the tests and also the travis file
Might be good also to specify somewhere the decoder it is compatible with (I guess just the rnn one).
Thanks.

**Comment by SteveKommrusch** [comment]
@vince62s, the paper "Graph2Seq: Graph to Sequence Learning with Attention-based neural networks": https://arxiv.org/pdf/1804.00823.pdf discusses using such a system for structured query-response in NLP (such as an SQL query). Using the system for NLP requires creating or having a graph representation of the input (such as a diagrammed sentence with parts of speech and clauses broken out). Using the system for computer code analysis, as we do in the dataflow graph equivalence paper, is 'easier' in that compilers already process code using ASTs (abstract syntax trees), which provide useful information for an ML to learn from.

I will work to clean up the files and add a docs/source.

You are correct that ggnn_encoder.py currently assumes an RNN decoder (indeed, I think only an LSTM decoder works).

**Comment by SteveKommrusch** [comment]
@vince62s I have added docs/source/ggnn.md (and reverted pull_request_chk). The ggnn.md contains instructions that use a new GitHub repo as an example; this same data can be used as a test, but it takes about 40 minutes to run on my particular system. What are the next steps towards approval?

**Comment by francoishernandez** [comment]
Hey @SteveKommrusch 
Thanks for this contribution!

I think we can merge this, though we may not promise to maintain it if there were any structural changes that would require it in the future. But if you stay around you may well do it!
Also, if some issues arise about this please keep an eye if we tag you.

40 minutes is a bit much... Isn't it possible to strip it to the bare minimum with only a few toy examples and a few steps?

**Comment by SteveKommrusch** [comment]
Thanks @francoishernandez
Instead of step 2 in the quickstart section of ggnn.md, I ran the command below and it completes in 4 minutes. The validation accuracy is only 54%, but that's good enough to check that the connections are still in place an no one broke the model (for robustness, any accuracy over 40% could be considered a passing case as results vary run-to-run). Should I just add this to ggnn.md, or is there a special place to put tests?

python train.py -data $data_path/final -encoder_type ggnn -layers 2 -decoder_type rnn -rnn_size 256 -learning_rate 0.1 -valid_steps 1000 -global_attention general -batch_size 32 -word_vec_size 256 -bridge -train_steps 1000 -gpu_ranks 0 -save_model $data_path/final-model -src_vocab $data_path/srcvocab.txt -n_edge_types 9 -state_dim 256 -n_steps 10 -n_node 64 > $data_path/train.final.out

**Comment by SteveKommrusch** [comment]
I should also note that I would expect to do some support to insure other users can benefit from my submission.

**Comment by francoishernandez** [comment]
Tests for Travis CI are defined in the [.travis.yml](https://github.com/OpenNMT/OpenNMT-py/blob/master/.travis.yml) file.
4 minutes is still a bit long for Travis (the whole run currently takes 4 to 5 minutes).
I think you can:
- keep the existing version in the docs -- for anyone wanting to reproduce it 40 minutes is decent;
- put an even shorter version (10 train_step for instance) in the .travis.yml -- just to check everything is running.

**Comment by SteveKommrusch** [comment]
I have added a 10-sample dataset with data/ggnn* and a new 10-step training test that runs in under a minute to .travis.yml.

**Comment by francoishernandez** [comment]
This seems much better, thanks!
I think you may just add a comment in `onmt.opts` to mention these new opts are for GGNN encoder related stuff.
And if you've got nothing else to change I think we can merge!

**Comment by SteveKommrusch** [comment]
I added comments to onmt/opts.py to clarify the new options related to 'encoder_type ggnn'.

**Comment by monperrus** [comment]
Nice, thanks Steve and François!
